diff --git a/misc-progs/inp.c b/misc-progs/inp.c
index 282d570cdbf5..a799632cef44 100644
--- a/misc-progs/inp.c
+++ b/misc-progs/inp.c
@@ -30,7 +30,6 @@
 #include <sys/types.h>
 #include <sys/stat.h>
 
-#include <asm/io.h> /* linux-specific */
 
 #ifdef __GLIBC__
 #  include <sys/perm.h>
diff --git a/misc-progs/outp.c b/misc-progs/outp.c
index 219d61307c9a..9e8a30d2d057 100644
--- a/misc-progs/outp.c
+++ b/misc-progs/outp.c
@@ -30,7 +30,6 @@
 #include <sys/types.h>
 #include <sys/stat.h>
 
-#include <asm/io.h> /* linux-specific */
 
 #ifdef __GLIBC__
 #  include <sys/perm.h>
diff --git a/misc-progs/setlevel.c b/misc-progs/setlevel.c
index fec666d9c5b3..683d7490392b 100644
--- a/misc-progs/setlevel.c
+++ b/misc-progs/setlevel.c
@@ -22,12 +22,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <errno.h>
-/* #include <unistd.h> */ /* conflicting on the alpha */
-#define __LIBRARY__ /* _syscall3 and friends are only available through this */
-#include <linux/unistd.h>
-
-/* define the system call, to override the library function */
-_syscall3(int, syslog, int, type, char *, bufp, int, len);
+#include <sys/klog.h>
 
 int main(int argc, char **argv)
 {
@@ -38,7 +33,7 @@ int main(int argc, char **argv)
     } else {
         fprintf(stderr, "%s: need a single arg\n",argv[0]); exit(1);
     }
-    if (syslog(8,NULL,level) < 0) {  
+    if (klogctl(8,NULL,level) < 0) {
         fprintf(stderr,"%s: syslog(setlevel): %s\n",
                 argv[0],strerror(errno));
         exit(1);
