diff --git a/drivers/md/dm-android-verity.c b/drivers/md/dm-android-verity.c
index c521df010ee3..eb4bdf66087c 100644
--- a/drivers/md/dm-android-verity.c
+++ b/drivers/md/dm-android-verity.c
@@ -643,6 +643,7 @@ static int add_as_linear_device(struct dm_target *ti, char *dev)
 	android_verity_target.status = dm_linear_status,
 	android_verity_target.prepare_ioctl = dm_linear_prepare_ioctl,
 	android_verity_target.iterate_devices = dm_linear_iterate_devices,
+        android_verity_target.direct_access = dm_linear_direct_access,
 	android_verity_target.io_hints = NULL;
 
 	set_disk_ro(dm_disk(dm_table_get_md(ti->table)), 0);
