diff --git a/misc-modules/jiq.c b/misc-modules/jiq.c
index 1c4cb565552f..dd9bbfd7b287 100644
--- a/misc-modules/jiq.c
+++ b/misc-modules/jiq.c
@@ -188,8 +188,10 @@ static struct timer_list jiq_timer;
 
 static void jiq_timedout(unsigned long ptr)
 {
-	jiq_print((void *)ptr);            /* print a line */
+	struct clientdata *data = (void*)ptr;
+	jiq_print((void *)data);            /* print a line */
 
+	data->wait_cond = 1;
 	wake_up_interruptible(&jiq_wait);  /* awake the process */
 }
 
@@ -202,7 +204,7 @@ static int jiq_read_run_timer_proc_show(struct seq_file *m, void *v)
 
 	init_timer(&jiq_timer);              /* init the timer structure */
 	jiq_timer.function = jiq_timedout;
-	jiq_timer.data = (unsigned long)m;
+	jiq_timer.data = (unsigned long)&jiq_data;
 	jiq_timer.expires = jiffies + HZ; /* one second */
 
 	jiq_print(&jiq_data);   /* print and go to sleep */
diff --git a/misc-modules/silly.c b/misc-modules/silly.c
index 3b1f8936471f..b708ecc840ae 100644
--- a/misc-modules/silly.c
+++ b/misc-modules/silly.c
@@ -77,7 +77,7 @@ enum silly_modes {M_8=0, M_16, M_32, M_memcpy};
 ssize_t silly_read(struct file *filp, char __user *buf, size_t count, loff_t *f_pos)
 {
 	int retval;
-	int mode = iminor(filp->f_dentry->d_inode);
+	int mode = iminor(filp->f_path.dentry->d_inode);
 	void __iomem *add;
 	unsigned long isa_addr = ISA_BASE + *f_pos;
 	unsigned char *kbuf, *ptr;
@@ -159,7 +159,7 @@ ssize_t silly_write(struct file *filp, const char __user *buf, size_t count,
 		    loff_t *f_pos)
 {
 	int retval;
-	int mode = iminor(filp->f_dentry->d_inode);
+	int mode = iminor(filp->f_path.dentry->d_inode);
 	unsigned long isa_addr = ISA_BASE + *f_pos;
 	unsigned char *kbuf, *ptr;
 	void __iomem *add;
