diff -U3 -r /home/zhenyanjie/myworkspace/seda/results/sdcardfs-diff/result-3.10-4.14/oldfile/sdcardfs-3.10/file.c /home/zhenyanjie/myworkspace/seda/results/sdcardfs-diff/result-3.10-4.14/newfile/sdcardfs-3.10/file.c
--- /home/zhenyanjie/myworkspace/seda/results/sdcardfs-diff/result-3.10-4.14/oldfile/sdcardfs-3.10/file.c	2018-04-16 00:35:02.103872366 +0800
+++ /home/zhenyanjie/myworkspace/seda/results/sdcardfs-diff/result-3.10-4.14/newfile/sdcardfs-3.10/file.c	2018-04-16 00:35:02.179873555 +0800
@@ -359,6 +359,7 @@
 
 
 const struct file_operations sdcardfs_main_fops = {
+	.get_unmapped_area = thp_get_unmapped_area,
 	.clone_file_range = cifs_clone_file_range,
 	.llseek		= generic_file_llseek,
 	.read		= sdcardfs_read,
@@ -377,6 +378,7 @@
 
 /* trimmed directory options */
 const struct file_operations sdcardfs_dir_fops = {
+	.get_unmapped_area = thp_get_unmapped_area,
 	.clone_file_range = cifs_clone_file_range,
 	.llseek		= sdcardfs_file_llseek,
 	.read		= generic_read_dir,
