diff --git a/sbull/sbull.c b/sbull/sbull.c
index 1a7d4ce3894a..ffab3c280f63 100644
--- a/sbull/sbull.c
+++ b/sbull/sbull.c
@@ -222,14 +222,13 @@ static void sbull_full_request(struct request_queue *q)
 /*
  * The direct make request version.
  */
-static int sbull_make_request(struct request_queue *q, struct bio *bio)
+static void sbull_make_request(struct request_queue *q, struct bio *bio)
 {
 	struct sbull_dev *dev = q->queuedata;
 	int status;
 
 	status = sbull_xfer_bio(dev, bio);
 	bio_endio(bio, status);
-	return 0;
 }
 
 
