diff --git a/sbull/sbull.c b/sbull/sbull.c
index d77ea2556bb8..5fcb25551311 100644
--- a/sbull/sbull.c
+++ b/sbull/sbull.c
@@ -201,13 +201,15 @@ static void sbull_full_request(struct request_queue *q)
 /*
  * The direct make request version.
  */
-static void sbull_make_request(struct request_queue *q, struct bio *bio)
+static blk_qc_t sbull_make_request(struct request_queue *q, struct bio *bio)
 {
 	struct sbull_dev *dev = q->queuedata;
 	int status;
 
 	status = sbull_xfer_bio(dev, bio);
-	bio_endio(bio, status);
+	bio->bi_error = status;
+	bio_endio(bio);
+	return BLK_QC_T_NONE;
 }
 
 
